lib/teleterm/gateway/gateway.go

```diff
@@ -32,19 +32,46 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// New creates an instance of Gateway
+// New creates an instance of Gateway. It starts a listener on the specified port but it doesn't
+// start the proxy – that's the job of Serve.
 func New(cfg Config) (*Gateway, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%s", cfg.LocalAddress, cfg.LocalPort))
+	result, err := newListenerAndLocalProxy(cfg, cfg.LocalPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cfg.LocalPort = result.LocalPort
+
+	gateway := &Gateway{
+		cfg:          &cfg,
+		closeContext: result.closeContext,
+		closeCancel:  result.closeCancel,
+		localProxy:   result.localProxy,
+	}
+
+	return gateway, nil
+}
+
+type newListenerAndLocalProxyResult struct {
+	LocalPort    string
+	listener     net.Listener
+	closeContext context.Context
+	closeCancel  context.CancelFunc
+	localProxy   *alpn.LocalProxy
+}
+
+func newListenerAndLocalProxy(cfg Config, port string) (*newListenerAndLocalProxyResult, error) {
```

Comment:
I made it so that functions which don't mutate `Config` accept a value rather than a pointer.